provider/packet/environ.go

```diff
@@ -0,0 +1,519 @@
+
+func (e *environ) AdoptResources(ctx context.ProviderCallContext, controllerUUID string, fromVersion version.Number) error {
+	return nil
+}
+
+func (e *environ) Bootstrap(ctx environs.BootstrapContext, callCtx context.ProviderCallContext, args environs.BootstrapParams) (*environs.BootstrapResult, error) {
+	return common.Bootstrap(ctx, e, callCtx, args)
+}
+
+func (e *environ) AllInstances(ctx context.ProviderCallContext) ([]instances.Instance, error) {
+	return e.getPacketInstancesByTag(map[string]string{"juju-model-uuid": e.Config().UUID()})
+}
+
+// if values tag and state are left empty it will return all instances
+func (e *environ) getPacketInstancesByTag(tags map[string]string) ([]instances.Instance, error) {
+	toReturn := []instances.Instance{}
+	packetTags := []string{}
+	for k, v := range tags {
+		packetTags = append(packetTags, fmt.Sprintf("%s=%s", k, v))
+	}
+
+	devices, _, err := e.packetClient.Devices.List(e.cloud.Credential.Attributes()["project-id"], nil)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, d := range devices {
+		cp := d
+		if isListContained(packetTags, cp) {
+			toReturn = append(toReturn, &packetDevice{e, &cp})
+		}
+	}
+
+	return toReturn, nil
+}
+
+func isListContained(tags []string, d packngo.Device) bool {
+	for _, t := range tags {
+		tagFound := false
+		for _, tt := range d.Tags {
+			if t == tt {
+				tagFound = true
+				break
+			}
+		}
+		if tagFound == false {
+			return false
+		}
+	}
+	return true
+}
+
+func (e *environ) AllRunningInstances(ctx context.ProviderCallContext) ([]instances.Instance, error) {
+	return e.getPacketInstancesByTag(map[string]string{"juju-model-uuid": e.Config().UUID()})
+}
+
+func (e *environ) Config() *config.Config {
+	e.ecfgMutex.Lock()
+	defer e.ecfgMutex.Unlock()
+	return e.ecfg.config
+}
+
+func (e *environ) ConstraintsValidator(ctx context.ProviderCallContext) (constraints.Validator, error) {
+	validator := constraints.NewValidator()
+	validator.RegisterUnsupported([]string{constraints.CpuPower, constraints.VirtType})
+	validator.RegisterConflicts([]string{constraints.InstanceType}, []string{constraints.Mem})
+	validator.RegisterVocabulary(constraints.Arch, []string{arch.AMD64, arch.ARM64, arch.I386, arch.PPC64EL})
+	return validator, nil
+}
+
+func (e *environ) ControllerInstances(ctx context.ProviderCallContext, controllerUUID string) ([]instance.Id, error) {
+	insts, err := e.getPacketInstancesByTag(map[string]string{"juju-is-controller": "true", "juju-controller-uuid": controllerUUID})
+	if err != nil {
+		return nil, err
+	}
+	instanceIDs := []instance.Id{}
+	for _, i := range insts {
+		instanceIDs = append(instanceIDs, i.Id())
+	}
+	return nil, nil
+}
+
+func (e *environ) Create(ctx context.ProviderCallContext, args environs.CreateParams) error {
+	return nil
+}
+
+func (e *environ) Destroy(ctx context.ProviderCallContext) error {
+	insts, err := e.getPacketInstancesByTag(map[string]string{"juju-model-uuid=": e.Config().UUID()})
```

Comment:
```suggestion
	insts, err := e.getPacketInstancesByTag(map[string]string{"juju-model-uuid": e.Config().UUID()})
```

There is a typo (redundant '=') in the key name which will cause the selection query to return no instances.